### ISSUE TITLE
Update and test neighbourhood import tools with Friesland, Netherlands data

### ISSUE THREAD

**Issue by stepps00** [whosonfirst-data/whosonfirst-data#870]
Before a PR for https://github.com/whosonfirst-data/whosonfirst-data/issues/837 is merged, the neighbourhood import tools will need updates to handle `localadmin` features.

**Comment by stepps00** [comment]
Friesland records are in the S3 `misc/Issue 870` folder.

New filenames include:

```
wof_localadmin.geojson
no_wof_localadmin.geojson
wof_localadmin_alt.geojson
no_wof_localadmin_alt.geojson
```

**Comment by thisisaaronland** [comment]
What exactly is going on here, with all these files?

What I am seeing is new compared the way we've always handled the neighbourhood imports. Specifically, we've always assumed a single parent locality for which its _descendants_ are being updated.

The files in 870 appear to be a mix of all the things - localities, new localadmins, localadmins with alt geometries (better suited for a secondary update or PR?), new localities (again this has not ever been part of the mix) and then finally neighbourhoods and their immediate relations.

I am not opposed to updating the neighbourhood import/update code but what I am seeing introduces a whole new set of requirements that I am not sure we have fully articulated.

**Comment by stepps00** [comment]
```
Specifically, we've always assumed a single parent locality for which 
its descendants are being updated.
```

Yeah, in the case of the Netherlands, we have many locality records that we've bundled into a single GeoJSON. I can see how this would be an issue with the current tools... right now, the neighbourhood tool is specific to updating records for a _single_ locality placetype and below.

The deliverable for Friesland is assuming a single parent region for which its descendants are being updated.

With the current tools, we would need to create a zip file for each locality being updated (nearly every locality in the Netherlands), then handle localadmin updates separately. I'm not sure that is the most efficient way to import these updates..

The thought is that we would have twelve deliverables for the Netherlands, one per region. I am open to splitting up the deliverables to make the imports more manageable, thoughts?

**Notes about the Friesland files:**

At the locality placetype, we're both updating locality records and introducing new locality records. I don't believe the `no_wof_locality.geojson` file has not been used in previous updates; the description of this file is: 

```
no_wof_locality.geojson - A single polygon feature that should be imported
 as a new default geometry for the given locality. In addition to any 
source-specific properties that we’ve kept from the source data files 
(stored in the source_data_aliases.json file), properties should map to 
wof-properties in the property_aliases.json file. Note: this file will likely 
never be created, but for the sake of including all possible filenames, 
it’s been added.
```

At the localadmin placetype, we're updating localadmin records and introducing new localadmin records. This includes alt records.

**Comment by nvkelso** [comment]
**Some background:** This Netherlands work is a big experiment. We started with a handful of locality updates using the "per locality & neighrbourhood descendants, adjacent neighbours" workflow that have already been merged. Reviewing that work it turned out the country had a pretty good set of neighbourhoods for the entire country that we were sourcing again and again. The work that @ZBSingleton is engaged with now is looking at using those, but we're rebuilding the entire lower-end of the hierarchy along the way to get better locality shapes as our existing Netherlands localities weren't very good or comprehensive. 

Should we go thru traditional PR process or BI neighbourhood pipeline for these 12 regions worth of Netherlands updates?

**Comment by thisisaaronland** [comment]
My thoughts are this:

1. Anything is possible with enough time and typing
2. My concerns are that this is a whole bunch of new process and potentially monkey-patching a tool that already works well, for an established task. Also creating sprawling logic-trains that are then difficult to understand and tease apart in to their component parts. I would like the emphasis to always be on small discrete pieces of functionality that can be run independently of one another and then stacked together (as it is the `neighbourhood-tool` has or is about to fail this test).
3. Generally the size of updates. Specifically, the larger the geographic area for an update (granted the Netherlands are the size of an average American pizza...) the longer it will take for any one task to complete and more importantly to spot problems or gotchas and the more difficult it will be to process things in parallel.

I am happy to try @stepps00 's proposal to see how/what works but I am still not sure I understand what or how to process the `_alt.geojson` files. Are these just alt files that need to be appended to a given record? If so this seems like a good candidate for a new and different task and should _not_ be an "oh, and if you see..." side-car to the process of rebuilding neighbourhoods.

**Comment by thisisaaronland** [comment]
And when I say "I am happy to try @stepps00 's proposal..." what I mean to say is: I am happy try @stepps00 's proposal pending a sanity-check on https://github.com/whosonfirst-data/whosonfirst-data/compare/issue-870

This should account for all the new files _except_ the alt files for reasons discussed above and because I wanted to see if the existing tools could be tweaked more or less as-is, without needing to add new code (for the stand-alone alt files).

**Comment by stepps00** [comment]
```
I am happy to try @stepps00 's proposal to see how/what works but I am still not sure I understand what or how to process the _alt.geojson files. Are these just alt files that need to be appended to a given record?
```

Yep - these alt records would be appended (via `src:alt_geom` prop) to the main record. No PIPing for these records.

And to you points 2 and 3, I agree that the current tooling works well to handle locality-specific updates, as it should. If we decide to create new tool or update existing logic, I am happy to split up the files and/or file a related PR, if needed.

Will take a look at the `issue-870` diff shortly...

**Comment by thisisaaronland** [comment]
Okay - It's not impossible to imagine adding the code to handle the `_alt` files but in the interest of making sure the other stuff works I am going to hold off pending a sanity check on the existing `870` changes.

**Comment by thisisaaronland** [comment]
These updates have exposed inadequacies in the hierarchy-related code where certain existential flags are not being trapped. Specifically, non-null `edtf:cessation` properties. For example:

```
DEBUG:root:[101806091][Balk] possible reverse parents : localadmin;county;region
DEBUG:root:SELECT id, parent_id, placetype_id, meta, ST_AsGeoJSON(geom), ST_AsGeoJSON(centroid) FROM whosonfirst WHERE ST_Intersects(geom, ST_GeomFromGeoJSON(%s)) AND is_superseded=%s AND is_deprecated=%s AND placetype_id=%s
DEBUG:root:time to execute query (point-in-polygon for 52.895805,5.583767) : 0.087082862854
DEBUG:root:find parent (localadmin) for 52.895805, 5.583767 : 2
```

There are two ways to address, and we're going to try both in the short-term:

1. Update all the PGIS related code/infrastructure to handle existential flags. This is not overly complicated but it does require non-zero time-and-typing (and debugging) to complete

* https://github.com/whosonfirst/go-whosonfirst-pgis/issues/4
* https://github.com/whosonfirst/go-whosonfirst-pgis/issues/5

2. Apply a band-aid to the `py-mz-wof-spatial/hierarchy` code to filter records with `edtf:cessation` properties post-PGIS

* https://github.com/whosonfirst/py-mapzen-whosonfirst-spatial/compare/is_ceased
* https://github.com/whosonfirst/py-mapzen-whosonfirst-hierarchy/compare/is_ceased

There is another task processing the records per (2) so I will check it in the morning. Assuming it works that should mean that (1) doesn't block updates in the short-term.

**Comment by thisisaaronland** [comment]
https://github.com/whosonfirst/py-mapzen-whosonfirst-spatial/commit/452d4a6eb993c2cb1f05920f27da3335c2dddeb8 is what happens when you write code at the end of the day.

Trying again...

**Comment by thisisaaronland** [comment]
Well, we'll need to re-index the PGIS store since it now has records of newly created descendants that get invoked during a PIP/intersects query (for their ancestor) before their record is written to disk.

There are a couple of approaches that _might_ mitigate re-indexing but it's also possible that re-indexing might simply be the fastest thing in the moment...

**Comment by thisisaaronland** [comment]
Okay... there is experimental code to "pre export" new files so they should be available during an ancestor/descendant rebuild.

This is experimental and a stop-gap to keep things moving and the "pre export" code will almost certainly be removed before hilarity ensues...

**Comment by thisisaaronland** [comment]
Try, try again... https://github.com/whosonfirst/py-mapzen-whosonfirst-spatial/commit/b7b8da6d433b7d74103c9e97961b4926c4330431

**Comment by stepps00** [comment]
Friesland data has been imported, other issues (https://github.com/whosonfirst-data/whosonfirst-data/issues/837, https://github.com/whosonfirst-data/whosonfirst-data/issues/931, https://github.com/whosonfirst-data/whosonfirst-data/issues/929) have been opened as follow-ups for the Netherlands.
